packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts

```diff
@@ -0,0 +1,109 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ApiKey, AppName, Auth } from '../../model/auth';
+import { User } from '../../model/user';
+import { PersistedBlob, Persistence } from '../persistence';
+import { UserImpl } from '../user/user_impl';
+import { inMemoryPersistence } from './in_memory';
+
+export const _AUTH_USER_KEY_NAME = 'authUser';
+export const _PERSISTENCE_KEY_NAME = 'persistence';
+const _PERSISTENCE_NAMESPACE = 'firebase';
+
+export function persistenceKeyName_(
+  key: string,
+  apiKey: ApiKey,
+  appName: AppName
+): string {
+  return `${_PERSISTENCE_NAMESPACE}:${key}:${apiKey}:${appName}`;
+}
+
+export class PersistenceUserManager {
+  private readonly fullUserKey: string;
+  private readonly fullPersistenceKey: string;
+  private constructor(
+    public persistence: Persistence,
+    private readonly auth: Auth,
+    private readonly userKey: string
+  ) {
+    const { config, name } = this.auth;
+    this.fullUserKey = persistenceKeyName_(this.userKey, config.apiKey, name);
+    this.fullPersistenceKey = persistenceKeyName_(
+      _PERSISTENCE_KEY_NAME,
+      config.apiKey,
+      name
+    );
+  }
+
+  setCurrentUser(user: User): Promise<void> {
+    return this.persistence.set(this.fullUserKey, user.toPlainObject());
+  }
+
+  async getCurrentUser(): Promise<User | null> {
+    const blob = await this.persistence.get<PersistedBlob>(this.fullUserKey);
+    return blob ? UserImpl.fromPlainObject(this.auth, blob) : null;
+  }
+
+  removeCurrentUser(): Promise<void> {
+    return this.persistence.remove(this.fullUserKey);
+  }
+
+  savePersistenceForRedirect(): Promise<void> {
+    return this.persistence.set(this.fullPersistenceKey, this.persistence.type);
+  }
+
+  async setPersistence(newPersistence: Persistence): Promise<void> {
+    if (this.persistence.type === newPersistence.type) {
+      return;
+    }
+
+    const currentUser = await this.getCurrentUser();
+    await this.removeCurrentUser();
+
+    this.persistence = newPersistence;
+
+    if (currentUser) {
+      return this.setCurrentUser(currentUser);
+    }
+  }
+
+  static async create(
+    auth: Auth,
+    persistenceHierarchy: Persistence[],
+    userKey = _AUTH_USER_KEY_NAME
+  ): Promise<PersistenceUserManager> {
+    if (!persistenceHierarchy.length) {
+      return new PersistenceUserManager(inMemoryPersistence, auth, userKey);
+    }
+
+    const key = persistenceKeyName_(userKey, auth.config.apiKey, auth.name);
+    for (const persistence of persistenceHierarchy) {
+      if (await persistence.get(key)) {
+        return new PersistenceUserManager(persistence, auth, userKey);
+      }
+    }
+
+    // Check all the available storage options.
+    // TODO: Migrate from local storage to indexedDB
```

Comment:
we should add these to the sheet so we don't forget

Reply:
Done